### ISSUE TITLE
Not all output from PostgreSQL console being shown

### ISSUE THREAD

**Issue by orthomind** [go-task/task#114]
- Task version: Git repo around May 23rd
- OS: Linux
- Local psql: 10.3

I've tried creating a simple task command to connect to a PostgreSQL instance.  Now, this works when I connect via the psql command via the docker container, but not when I use my locally installed psql client:

```
version: "2"

tasks:
  pg_start:
    desc: Starts a docker instance of PostgreSQL
    cmds:
      - docker run -p 5432:5432 --rm --name="testps" -e "POSTGRES_PASSWORD=somepass" -e "POSTGRES_USER=someuser" -e "POSTGRES_DB=somedb" -d postgres:10.3
  pg_console_local:
    desc: Connects to console via local psql
    cmds:
      - PGPASSWORD=somepass psql -h localhost -p 5432 -U someuser somedb
  pg_console_docker:
    desc: Connects to console via docker psql
    cmds:
      - docker exec -ti -e "PGPASSWORD=somepass" testps psql -h localhost -p 5432 -U someuser somedb
```

Output from this (note that with the pg_console_local attempt I could use \q to quit, but no prompt was shown to indicate I could enter commands):

```
$ task pg_start
docker run -p 5432:5432 --rm --name="testps" -e "POSTGRES_PASSWORD=somepass" -e "POSTGRES_USER=someuser" -e "POSTGRES_DB=somedb" -d postgres:10.4
fe2247e778fc7146380fbd4f0ffce7b997a14f02aa13da1d8d0df156e2f1a1d2

$ task pg_console_docker
docker exec -ti -e "PGPASSWORD=somepass" testps psql -h localhost -p 5432 -U someuser somedb
psql (10.4 (Debian 10.4-2.pgdg90+1))
Type "help" for help.

somedb=# \q

$ task pg_console_local 
PGPASSWORD=somepass psql -h localhost -p 5432 -U someuser somedb
\q

$
```

**Comment by smyrman** [comment]
If you run docker in tty mode (`-t`), I suspect you might see the same output in docker as locally.

I would assume the postgreSQL prompt is not printed to stdout or stderr, but to /dev/tty:
https://unix.stackexchange.com/questions/20826/which-stream-does-bash-write-its-prompt-to

Task and/or the shell implementation task rely on is probably does not set to forward tty output from sub-processes (task commands).

**Comment by andreynering** [comment]
Hi @orthomind,

Thanks for reporting the issue! I'll investigate this when having a bit of available time.

@smyrman Thanks for the input, I think you're probably right.

**Comment by andreynering** [comment]
Hello guys.

Some more info. I just investigated this issue a bit. Docker is not required to reproduce it, it also happens if you just call `psql ...` directly in the Taskfile.

I tried some interactive terminal applications, like `vim`, [`micro`](https://github.com/zyedidia/micro), others and seems that most apps fallback to `stdin`/`stdout` when PTY is not available and just work as expected.

`psql` seems to be an exception. Maybe this can be considered bug on `psql` itself (i.e. not detecting PTY/fallbacking to `stdin`/`stdout`).

I tried implementing PTY redirection on [mvdan/sh](https://github.com/mvdan/sh) using [kr/pty](https://github.com/kr/pty) but the result was buggy, and I don't want to push a half-working solution.

TL;DR: There's no easy fix for this. Let's keep this issue open for now.

**Comment by GuillaumeAmat** [comment]
Hi there, I can see that Tusk does not suffer from that issues. And AFAIK, it is also coded in Go. Maybe there is some inspiration to take there?

**Comment by smyrman** [comment]
I believe `tusk` rely on the OS shell for interpretation:
- https://github.com/rliebz/tusk/blob/00271e09bc9ce6735200fc4e9a46b52e68909f16/config/task/exec.go

This allows the TTY / PTY redirection to be handled by the shell command. However, it also makes task files less portable, as running a task on a differently configured system, or different platform (Linux v.s. Mac OS v.s. Windows) could give very different results.

go-task emebes a shell interpreter (mvdan/sh). This introduce some bugs as mvdan/sh is naturaly less mature than the local shell implementations, but all bugs (and intended behavior) is consistent across systems, and that is a lot more valuable, I think.

**Comment by GuillaumeAmat** [comment]
I totally understand and support your point @smyrman. But in the current state, some commands are completely unusable. That let us in a strange state where only a part of the project commands are available in the Taskfile.

Could you consider to add an option like `--os-shell`, which could use the os native shell to handle the redirections?

**Comment by smyrman** [comment]
Hopefully it will eventually be possible to solve it in mvdan/sh, but it would require someone to do the work.

What you can try as a work-around @GuillaumeAmat, is perhaps something like this:

```ymal
version: 2
tasks:
  mytask:
    cmds:
    - sh -c '<COMMAND>' 
```

PS! haven't tried this myself.

**Comment by GuillaumeAmat** [comment]
> Hopefully it will eventually be possible to solve it in mvdan/sh, but it would require someone to do the work.

Hope so too. Maybe with a revival of that issue: https://github.com/mvdan/sh/issues/247

> What you can try as a work-around @GuillaumeAmat, is perhaps something like this:
> 
> ```
> version: 2
> tasks:
>   mytask:
>     cmds:
>     - sh -c '<COMMAND>' 
> ```
> 
> PS! haven't tried this myself.

Eventually, the command is launched with `mvdan/sh` so the result is the same. Thanks for the idea anyway :)

**Comment by andreynering** [comment]
Hi @GuillaumeAmat,

Yeah, I know this issue is pretty important, even being almost a year since this was reported.

I haven't invested more time on this recently, but I think I should take another look on this soon.

It's interesting that Tusk doesn't suffer from this problem, despite not implementing PTY. I definitely need to try running some stuff on raw Go vs. mvdan/sh to understand better where the problem is.

What tools are you trying to use that are also suffering from this issue (other than psql)?

Also, help with debugging and eventual contributions to Task or mvdan/sh are more than welcome.

**Comment by GuillaumeAmat** [comment]
> I haven't invested more time on this recently, but I think I should take another look on this soon.

Great thanks!

I'm a web developer and I use `jest` and `yarn` a lot. Both have interactive commands that:

* don't display everything: `jest --watch`
* refuse completely to run: `yarn upgrade-interactive`

**Comment by andreynering** [comment]
Hi,

I think some (hopefully most) issues similar to the ones described here were just fixed on master after #200. (I tested `ruby` and `psql` and they're now working fine).

Please, try master with the tools you've experienced this issue and let me know if it was fixed on not.

Thanks!

/cc @orthomind @GuillaumeAmat

**Comment by GuillaumeAmat** [comment]
Great to read @andreynering! 

I'm not familiar to Go development though, and just being a new dad (hence the answer delay ;) ), so I don't have much time to learn how to install/build the project locally. Can you give me some commands I can copy/paste to test `master`? It would be much appreciated!

**Comment by andreynering** [comment]
Hi @GuillaumeAmat,

First of all, congrats! 🙂

I did better. I just released the fix, so you just need to upgrade using your install method of choice.

**Comment by GuillaumeAmat** [comment]
Thanks @andreynering ;)

The new release works perfectly! Thank you very much for your great project, it is so delightful to work with it everyday.
